mod/widgets/set/self/jurisdiction.rb

```diff
@@ -1,11 +1,63 @@
 format :json do
   view :select2 do
-    { results: option_list }.to_json
+    { results: select2_option_list }.to_json
   end
 
-  def option_list
-    Card.search(type_id: JurisdictionID).each_with_object([]) do |i, ar|
-      ar << { id: i.key, text: i.name }
+  def select2_option_list
+    if name_query
+      wql = { type_id: JurisdictionID,
+              name:  ["match", name_query] }
+      Card.search(wql).each_with_object([]) do |i, ar|
+        ar << { id: i.codename, text: i.name }
+      end
+    else
+      group_by_country
     end
   end
+
+  # The jurisdiction names from OpenCorporates sometimes have
+  # the country of a state in brackets like "California (United States)".
+  # We group these by country. But there are also
+  # entries that have a clarification in brackets like "Holy See (Vatican City State)"
+  # So if a group has only one child at the end we remove the group and put the "country"
+  # back into brackets
+  def group_by_country
+    h = Hash.new { |hash, key| hash[key] = {} }
+    by_country = Card.search(type_id: JurisdictionID).each_with_object(h) do |i, groups|
+      if (m = i.name.match(/(?<state>.+?)\s*\((?<country>[^)]+)\)/))
+        groups[m[:country]][:text] ||= m[:country]
+        groups[m[:country]][:children] ||= []
+        groups[m[:country]][:children] << { id: i.codename, text: m[:state] }
+      else
+        groups[i.name][:text] ||= i.name
+        groups[i.name][:id] = i.codename
+      end
```

Comment:
this would be a good candidate for multiple methods.
```
if (m = i.name.match REGEXPCONSTANT)
  country_details groups[m[:country]], i, m
else
  group_details groups[:i.name], i
end
```